plugins/jabra-file/fu-jabra-file-device.c

```diff
@@ -0,0 +1,855 @@
+				  FU_JABRA_FILE_RETRY_DELAY,
+				  (gpointer)&rx_data,
+				  error))
+		return FALSE;
+	version = fu_memstrsafe(rx_data.rxbuf,
+				sizeof(rx_data.rxbuf),
+				0x8,
+				sizeof(rx_data.rxbuf) - 8,
+				error);
+	if (version == NULL)
+		return FALSE;
+	fu_device_set_version(FU_DEVICE(self), version);
+	return TRUE;
+}
+
+static gboolean
+fu_jabra_file_device_file_checksum(FuJabraFileDevice *self,
+				   gchar *firmware_checksum,
+				   gboolean *match,
+				   GError **error)
+{
+	guint8 info = 0x01 << 6;
+	guint8 device_checksum[16] = {0x00};
+	g_autofree gchar *device_checksum_formatted = NULL;
+
+	FuJabraFileTxData tx_data1 = {
+	    .txbuf =
+		{
+		    FU_JABRA_FILE_IFACE,
+		    self->address,
+		    0x00,
+		    self->sequence_number,
+		    0x47,
+		    0x03,
+		    0x03,
+		    0x00,
+		},
+	    .timeout = FU_JABRA_FILE_STANDARD_SEND_TIMEOUT,
+	};
+	FuJabraFileTxData tx_data2 = {
+	    .txbuf =
+		{
+		    FU_JABRA_FILE_IFACE,
+		    self->address,
+		    0x00,
+		    (self->sequence_number) + 1,
+		    0x47,
+		    0x03,
+		    0x03,
+		    info,
+		},
+	    .timeout = FU_JABRA_FILE_STANDARD_SEND_TIMEOUT,
+	};
+	FuJabraFileRxData rx_data = {
+	    .rxbuf = {0x00},
+	    .timeout = FU_JABRA_FILE_STANDARD_RECEIVE_TIMEOUT,
+	};
+	g_autofree gchar *version = NULL;
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_jabra_file_device_tx_cb,
+				  FU_JABRA_FILE_MAX_RETRIES,
+				  FU_JABRA_FILE_RETRY_DELAY,
+				  (gpointer)&tx_data1,
+				  error))
+		return FALSE;
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_jabra_file_device_rx_with_sequence_cb,
+				  FU_JABRA_FILE_MAX_RETRIES,
+				  FU_JABRA_FILE_RETRY_DELAY,
+				  (gpointer)&rx_data,
+				  error))
+		return FALSE;
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_jabra_file_device_tx_cb,
+				  FU_JABRA_FILE_MAX_RETRIES,
+				  FU_JABRA_FILE_RETRY_DELAY,
+				  (gpointer)&tx_data2,
+				  error))
+		return FALSE;
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_jabra_file_device_rx_with_sequence_cb,
+				  FU_JABRA_FILE_MAX_RETRIES,
+				  FU_JABRA_FILE_RETRY_DELAY,
+				  (gpointer)&rx_data,
+				  error))
+		return FALSE;
+	if (rx_data.rxbuf[5] == 0xFE) {
+		*match = FALSE;
+		return TRUE;
+	}
+	if (!fu_memcpy_safe(device_checksum,
+			    16,
+			    0,
+			    rx_data.rxbuf,
+			    FU_JABRA_FILE_BUF_SIZE,
+			    12,
+			    16,
+			    error))
+		return FALSE;
+	device_checksum_formatted =
+	    g_strdup_printf("%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x",
+			    device_checksum[0],
+			    device_checksum[1],
+			    device_checksum[2],
+			    device_checksum[3],
+			    device_checksum[4],
+			    device_checksum[5],
+			    device_checksum[6],
+			    device_checksum[7],
+			    device_checksum[8],
+			    device_checksum[9],
+			    device_checksum[10],
+			    device_checksum[11],
+			    device_checksum[12],
+			    device_checksum[13],
+			    device_checksum[14],
+			    device_checksum[15]);
```

Comment:
can we do this in a loop with g_string_append_printf or fu_byte_array_to_string() :)

Reply:
definitely! I just couldn't find a smarter way when coding this, so I just went the good old printf, hoping for a good suggestion.